test/unittest/tests/test_acquire_release_macros.cpp

```diff
@@ -0,0 +1,80 @@
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include <catch2/catch.hpp>
+
+#include "wt_internal.h"
+
+#define TEST_ACQUIRE_TYPE(type)                      \
+    TEST_CASE("Test acquiring a " #type, "[acqrel]") \
+    {                                                \
+        type a = 5;                                  \
+        type *ap = &a;                               \
+        type a_result;                               \
+        WT_ACQUIRE_READ(a_result, *ap);              \
+        REQUIRE(a == a_result);                      \
+                                                     \
+        type b;                                      \
+        type *bp = &b;                               \
+        type b_result = 5;                           \
+        WT_ACQUIRE_READ_WITH_BARRIER(b_result, *bp); \
+        REQUIRE(b == b_result);                      \
+    }
+
+#define TEST_RELEASE_TYPE(type, value)               \
+    TEST_CASE("Test releasing a " #type, "[acqrel]") \
+    {                                                \
+        type a;                                      \
+        type *ap = &a;                               \
+        type a_set = value;                          \
+        WT_RELEASE_WRITE(*ap, a_set);                \
+        REQUIRE(a == a_set);                         \
+                                                     \
+        type b;                                      \
+        type *bp = &b;                               \
+        type b_set = value;                          \
+                                                     \
+        WT_RELEASE_WRITE_WITH_BARRIER(*bp, b_set);   \
+        REQUIRE(b == b_set);                         \
+    }
+
+/*
+ * We can't test that the specific instructions we want are being output, but having written this
+ * test we found that clang failed to compile for sizes that weren't uint64_t without some
+ * adjustments. So by compiling this test we also check that compilation works.
+ *
+ * This also helps verify the type checking in the acquire read macro, and that the barrier version
+ * doesn't produce different results.
+ */
+TEST_ACQUIRE_TYPE(uint64_t);
+TEST_ACQUIRE_TYPE(uint32_t);
+TEST_ACQUIRE_TYPE(uint16_t);
+TEST_ACQUIRE_TYPE(uint8_t);
+
+/* Test each branch of the release macro. */
+TEST_RELEASE_TYPE(uint64_t, UINT32_MAX + 1);
+TEST_RELEASE_TYPE(uint32_t, UINT16_MAX + 1);
+TEST_RELEASE_TYPE(uint16_t, UINT8_MAX + 1);
+TEST_RELEASE_TYPE(uint8_t, 1);
+
+/*
+ * We prevent users from releasing a value which isn't the same size as the type being released to.
+ * If unchecked it could result in integer overflow, both signed and unsigned. However this creates
+ * an issue when hash defined values are released, as they are assumed to be 8 bytes. We can work
+ * around this by casting in the hash define.
+ */
+TEST_CASE("Demonstrate hash define int size workaround", "[acqrel]")
```

Comment:
Does this test fail if we don't cast TEST_VALUE? I'm also curious why we cast to uint when `a` is a signed int.

Reply:
Additionally what does this mean for `#define`d values that don't have a typecast like
```
#define WT_PREPARE_INIT 0
#define WT_PREPARE_INPROGRESS 1
#define WT_PREPARE_LOCKED 2
#define WT_PREPARE_RESOLVED 3
```
?

Reply:
When we convert `WT_RELEASE_WRITE_WITH_BARRIER` to `WT_RELEASE_WRITE` we'll need to do the cast method. These #defines were what lead to me having to find that solution.
I added a comment to the test ec9d34273, explaining that it won't compile with the cast.

The cast may be ugly but it gives us correctness checking.

Reply:
We should keep this thread unresolved for others to see.

Reply:
SGTM. I like the idea of added proper type information to `#define WT_PREPARE_INIT 0` in a future ticket.